### ISSUE TITLE
terminal: `$PATH` has duplicated entries at the end (on macOS)

### ISSUE THREAD

**Issue by pantheraleo-7** [zed-industries/zed#21678]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

MRE:
- add a location to `$PATH`
- open the integrated terminal, and do `echo $PATH`

This location will be appended automatically, leading to two entries of each location in the `$PATH`.

This happens only in zed's integrated (alacritty?) terminal. I've checked iTerm2, Warp and Terminal.

### Environment

Zed: v0.165.3 (Zed Preview)
OS: macOS 15.1.1
Memory: 8 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

<img width="652" alt="Screenshot 2024-12-07 at 1 26 13 PM" src="https://github.com/user-attachments/assets/01354df2-7e4d-4ae9-bb2e-965f8e65fb92">
<img width="1466" alt="image" src="https://github.com/user-attachments/assets/10c347c0-0b39-4412-9441-9ae6fd5560c7">

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

```
[Zed.log](https://github.com/user-attachments/files/18047632/Zed.log)

```
</details>

**Comment by pantheraleo-7** [comment]
\cc @mrnugget

**Comment by mrnugget** [comment]
In your MRE, how do you launch Zed?

**Comment by pantheraleo-7** [comment]
GUI. CLI. It happens with both.

**Comment by mrnugget** [comment]
So that happens with an empty `.zshrc` and just that line in `.zprofile`?

**Comment by pantheraleo-7** [comment]
No, sorry. I miscommunicated. I put `$PATH` related thing in `.zprofile` and everything else in `.zshrc`, so I said what I said because of habit ig.

This is what is happening: whatever I add to the `$PATH` in `.zprofile` or `.zshrc` whether at the start or the end, it is geting appended to the `$PATH`.

So if I do `export PATH="$PATH:/hello/there/end"`, the `$PATH` is 
```console
/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/asadullahshaikh/.local/share/ds/bin:/opt/homebrew/bin:/Users/asadullahshaikh/.cargo/bin:/Users/asadullahshaikh/.local/bin:/hello/there/end:/hello/there/end
```

[.zprofile](https://github.com/user-attachments/files/18080641/zprofile.txt))
[.zshrc](https://github.com/user-attachments/files/18080645/zshrc.txt)

**Comment by mrnugget** [comment]
I can't reproduce this when I open Zed from the Dock on macOS:
![screenshot-2024-12-10-16 28 26@2x](https://github.com/user-attachments/assets/80014cea-c417-4dfe-9d98-8dffd5013669)

And when I launch it from the CLI I also can't reproduce it:

![screenshot-2024-12-10-16 30 31@2x](https://github.com/user-attachments/assets/a1572a81-1aaf-47bc-a9a2-711cb07669b6)

**Comment by pantheraleo-7** [comment]
Does zed modifies `$PATH` in anyway? I'm trying to debug this, and so far I can only confirm that this is _not_ a case of double sourcing of dotfiles (obv) and that it is a problem with zed particularly (I've tested iTerm2, Warp, Terminal, Alacritty) and it has something to do with `$PATH` manipulation.

**Comment by mrnugget** [comment]
It doesn't modify it. It only tries to get the `PATH` that's the most likely what the user expects. See here: https://zed.dev/docs/environment

We do try to be clever with the Python venv stuff though: https://zed.dev/docs/languages/python#virtual-environments

Do you have anything like that configured? Since I can't reproduce, I think there's a difference in our shell/Zed configuration and I wonder what that might be.

**Comment by pantheraleo-7** [comment]
> Do you have anything like that configured?

I don't understand what do you mean by this. I do have a venv on `$PATH`, I use [uv](https://github.com/astral-sh/uv) which is python project & package manager, but I don't think these interfere with zed and how it 'reads' `$PATH`.

As for configurations, I have shared my dotfiles in my second last reply and here's my [settings.json](https://github.com/user-attachments/files/18100768/settings.json) as well. Logs are attached in the issue write-up.

Do tell me what can I provide more. This is really bugging me.

I tried with **no** configurations at all i.e. an empty `.zshrc` and a `.zprofile` with only `$PATH` modifications and nothing else, I could still reproduce.

One more thing I've noticed is that the path which is added first in the dotfiles is appended last, like a stack.

**Comment by mrnugget** [comment]
Good news: I can reproduce it. I had `typeset -U PATH` in my `.zshrc` which ensures the values in `PATH` are unique.

Bad news: still haven't figured out why it happens.

**Comment by mrnugget** [comment]
Okay, I have it figured out. Here's what happens:

1. When Zed is launched from the Dock, we launch a login-shell to get the environment variables the user has configured in their shell environment, so we can better find language servers etc. On macOS, when you launch a GUI app, it essentially has no environment, so launching a login-shell and getting `env` is a common way to make a GUI app behave as if it was launched from the CLI. That happens here: https://github.com/zed-industries/zed/blob/59afc27f03705485d3701efeffa0e7cebc403916/crates/zed/src/main.rs#L1046-L1096
2. Zed takes that environment it gets from the login shell and sets it on itself (on the process.)
3. (When a login-shell is launched, both files — `.zprofile` and `.zshrc` — are sourced. `.zprofile` is sourced per login-shell, `.zshrc` per every shell instance.)
4. When a terminal is then launched, Alacritty _also_ spawns a login shell, which then inherits the environment of its parent process - Zed. Since that login shell _again_ sources `.zprofile`, the env is modified again and, in our case, the `PATH` gets the same value appended again.

So, shooting from the hip here, I'd say the "fix" would be to tell Alacritty to _not_ launch a login shell, since we effectively already have "a login-shell" in the form of the process environment — either when Zed is launched from the Dock (and we launched the login shell to get `env`) or when it was launched from CLI, in which case it inherits the CLI environment.

**Comment by mrnugget** [comment]
Workaround for now: set the `terminal.shell` program manually, so that Alacritty doesn't launch a login-shell.

With `terminal: { shell: {program: "zsh" } }` in `settings.json`:
![screenshot-2024-12-12-08 22 59@2x](https://github.com/user-attachments/assets/9791f8c6-b55b-4604-b11a-303a189e6ab7)

(Current state) with Alacritty using the system login shell again:
![screenshot-2024-12-12-08 24 31@2x](https://github.com/user-attachments/assets/99cbdfc7-779e-476b-8921-03692d99a5a8)

**Comment by mrnugget** [comment]
Looks like there's no easy option to tell Alacritty to not use a login shell. Only option is to set a program (which is what I suggested as "workaround")

https://github.com/alacritty/alacritty/blob/8731d157be4e49713587325f29099579d637a06a/alacritty_terminal/src/tty/unix.rs#L165-L183

**Comment by pantheraleo-7** [comment]
> So, shooting from the hip here, I'd say the "fix" would be to tell Alacritty to not launch a login shell

Yeah but `.zshrc` would still be sourced twice, wouldn't it? IMO the current implementation seems more like a hack. How does VS Code do it?

**Comment by pantheraleo-7** [comment]
@mrnugget will this be resolved?

**Comment by pantheraleo-7** [comment]
> it doesn't just seem a case of duplication

the core issue is the same (regardless of the end result) which is that the dotfiles are getting sourced multiple times. This was found and reproduced in this issue, a workaround was given by @mrnugget, but then github actions closed it due to inactivity.

> #22575#issuecomment

you might be able to solve your specific issue by changing the dotfile where you append to `$PATH` (meaning, try to append in `.zprofile` instead of `.zshrc`)
